### ISSUE TITLE
Ability to register custom regexes to be extracted

### ISSUE THREAD

**Issue by MartijnWelker** [MichaReiser/webpack-angular-translate#36]
In the grunt extraction plugin it's possible to [register your own regexes](https://github.com/angular-translate/grunt-angular-translate#customregex) to be extracted, are there any plans to also incorporate this into the webpack plugin ? 

If I understand the documentation correctly you could in theory just register all occurrences of the custom regexes using `i18n.registerTranslations` but this makes migrating from the grunt plugin to this plugin quite hard.

**Comment by MichaReiser** [comment]
Hi @MartijnWelker 

Currently, there are no plans to add such a feature since I'm no longer using angular myself. So at the moment I just ensure that the plugin works with the latest webpack versions but are not adding any new features.

Can you provide me with an actual use case for what you are using the custom regexes or what you want to achieve? The best is if you can post a snippet of your actual code and describe what the expected output should be.

**Comment by MartijnWelker** [comment]
Hi @MichaReiser, thank you for your quick response. I'll elaborate our use-case a bit further.
We sometimes call `$translate.instant` on a variable that's taken from an component attribute like so:

**Template**:
```
<rm-checkbox data-label="This will be the checkbox's label"></rm-checkbox>
```
**Component**:
```
this.labelTranslation = $translate.instant($attrs.label);
```


And then automatically extract the `data-label` with a custom regex in the grunt script

**Grunt config**
```
i18nextract: {
    default_options: {
        customRegex: [
            'rm-checkbox[^>]*?data-label="([^{]+?)"'
        ]
    }
}
```

When running the grunt script it will run all default regexes (to extract all translate filters/service calls) and all custom regexes and add the result of both to the output file.

**Comment by MichaReiser** [comment]
Hi @MartijnWelker 

Thank you for explaining your use case. The idea is to have the possibility to specify an element (tag-name) and the name of an attribute that behaves the same as the translate-attribute directive? The attribute value is the translation id and the default text?

**Comment by MartijnWelker** [comment]
Hi @MichaReiser.

I was thinking being able to hook into the loaders with a custom callback (maybe just like how [`matchAngularExpressions`](https://github.com/MichaReiser/webpack-angular-translate/blob/master/src/html/ng-filters.ts#L43) extracts a regex) would probably be the most flexible solution. If a specific return format is defined you're then free to create a function that can extract whatever you want.

For anyone facing the same issue; I've used the stand-alone version of the grunt script over at [angular-translate-extract](https://github.com/Boulangerie/angular-translate-extract), but this unfortunately has to be done outside of the webpack flow

**Comment by MichaReiser** [comment]
That sounds interesting (and extensible). An implementation of a callback that just extracts the value of an attribute could be provided by the plugin itself.

There are two related open questions

**How can I register the callback**

It's probably sufficient if the loader provides a mechanism to register a custom handler for a specific tag-name. In your case, you could register a callback for the rm-checkbox. Should we allow regular expressions to match the tag name as well?

**The signature of the callback**

If the callback is registered by tag name then I think the signature should be as follow:

```ts
(tagName: string, attributes: { [name: string]: string }) => Translation[]
```

where a `Translation` is defined as 

```
interface Translation {
  translationId: string;
  defaultText?: string
}
```

What do you think?

**Comment by MartijnWelker** [comment]
Sounds like a good idea! I think allowing a regex will be the most flexible solution because it also works for people using attribute directives instead of components.

**Comment by MichaReiser** [comment]
Hi @MartijnWelker 

I created a draft PR #38  addressing this issue. Could you verify if this approach would satisfy your needs and if you can help to clarify any of the open issues.

**Comment by MichaReiser** [comment]
#38  Released with version 3.2.0
